text_extensions_for_pandas/array/tensor.py

```diff
@@ -0,0 +1,223 @@
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+#
+# tensor.py
+#
+# Part of text_extensions_for_pandas
+#
+# Pandas extensions to support columns of N-dimensional tensors of equal shape.
+#
+
+from typing import *
+
+import numpy as np
+import pandas as pd
+from memoized_property import memoized_property
+
+# Internal imports
+import text_extensions_for_pandas.util as util
+
+
+@pd.api.extensions.register_extension_dtype
+class TensorType(pd.api.extensions.ExtensionDtype):
+    """
+    Panda datatype for a span that represents a range of characters within a
+    target string.
+    """
+
+    @property
+    def type(self):
+        # The type for a single row of a column of type CharSpan
+        return np.ndarray
+
+    @property
+    def name(self) -> str:
+        """A string representation of the dtype."""
+        return "Tensor"
+
+    @classmethod
+    def construct_from_string(cls, string: str):
+        """
+        See docstring in `ExtensionDType` class in `pandas/core/dtypes/base.py`
+        for information about this method.
+        """
+        # Upstream code uses exceptions as part of its normal control flow and
+        # will pass this method bogus class names.
+        if string == cls.__name__:
+            return cls()
+        else:
+            raise TypeError(
+                f"Cannot construct a '{cls.__name__}' from '{string}'")
+
+    @classmethod
+    def construct_array_type(cls):
+        """
+        See docstring in `ExtensionDType` class in `pandas/core/dtypes/base.py`
+        for information about this method.
+        """
+        return TensorArray
+
+
+class TensorArray(pd.api.extensions.ExtensionArray):
+    """
+    A Pandas `ExtensionArray` that represents a column of `numpy.ndarray`s,
+    or tensors, where the outer dimension is the count of tensors in the column.
+    """
+
+    def __init__(self, values: Union[np.ndarray, List[np.ndarray]],
+                 make_contiguous: bool = True):
+        """
+        :param values: A `numpy.ndarray` or list of `numpy.ndarray`s
+        :param make_contiguous: force values to be contiguous in memory
+        """
+        if isinstance(values, list):
+            self._tensor = np.stack(values, axis=0)
+        elif isinstance(values, np.ndarray):
+            self._tensor = values
+        else:
+            raise TypeError("Expected a numpy.ndarray or list of numpy.ndarray")
+        
+        if not self._tensor.flags.c_contiguous:
+            if make_contiguous:
+                self._tensor = np.ascontiguousarray(self._tensor)
+            else:
+                raise ValueError("Input ndarray is not contiguous")
```

Comment:
I don't think this branch is an error condition. The user has set `make_contiguous` to `False`, which means that the internal representation can be either contiguous or not.

Reply:
oh right, I think this was held over from the arrow extension, where it had to be contiguous